src/chunk new.rs

```diff
@@ -0,0 +1,289 @@
+                    data => Self::Unknown(data)
+                }
+            }
+        }
+
+        impl std::ops::Deref for ChunkType {
+            type Target = [u8; 4];
+
+            fn deref(&self) -> &Self::Target {
+                match self {
+                    $(Self::$chunk => &$chunk,)*
+                    Self::Unknown(ref data) => data
+                }
+            }
+        }
+
+        impl std::fmt::Debug for ChunkType {
+            fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+                match self {
+                    $(Self::$chunk => write!(f, "{:?}", $chunk),)*
+                    Self::Unknown(data) => f.debug_struct("ChunkType")
+                        .field("type", data)
+                        .field("critical", &is_critical(*data))
+                        .field("private", &is_private(*data))
+                        .field("reserved", &reserved_set(*data))
+                        .field("safecopy", &safe_to_copy(*data))
+                        .finish()
+                }
+            }
+        }
+
+        $(
+            $(#[doc=$desc])?
+            #[derive(PartialEq, Eq)]
+            pub struct $chunk;
+
+            impl Chunk for $chunk {
+                const TYPE: [u8; 4] = $type;
+                const MAX_LENGTH: usize = $size as usize;
+            }
+
+            impl std::ops::Deref for $chunk {
+                type Target = [u8; 4];
+
+                fn deref(&self) -> &Self::Target {
+                    &Self::TYPE
+                }
+            }
+
+            impl std::fmt::Debug for $chunk {
+                fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+                    f.debug_struct("ChunkType")
+                        .field("type", &stringify!($chunk))
+                        $(.field("description", &$desc))?
+                        .field("critical", &Self::is_critical())
+                        .field("private", &Self::is_private())
+                        .field("reserved", &Self::reserved_set())
+                        .field("safecopy", &Self::safe_to_copy())
+                        .finish()
+                }
+            }
+        )*
+    };
+}
+
+// -- Chunk type determination --
+
+/// Returns true if the chunk is critical.
+const fn is_critical(chunk: [u8; 4]) -> bool {
+    chunk[0] & 0b_00100000 == 0
+}
+
+/// Returns true if the chunk is private.
+const fn is_private(chunk: [u8; 4]) -> bool {
+    chunk[1] & 0b_00100000 != 0
+}
+
+/// Checks whether the reserved bit of the chunk name is set.
+/// If it is set the chunk name is invalid.
+const fn reserved_set(chunk: [u8; 4]) -> bool {
+    chunk[2] & 0b_00100000 != 0
+}
+
+/// Returns true if the chunk is safe to copy if unknown.
+const fn safe_to_copy(chunk: [u8; 4]) -> bool {
+    chunk[3] & 0b_00100000 != 0
+}
+
+use io::Write;
+use std::{io, ops};
+
+pub trait Chunk: ops::Deref<Target = [u8; 4]> {
```

Comment:
This whole trait is part of the overengineered feeling, traits are fancy but what kind of behaviour are we encapsulating here? Is there multiple ways to implement `is_private`? No, of course not. Don't misunderstand, I do appreciate the idea of a strong type for `Ihdr`, and especially its associated method, but this specific trait interface is rather odd. I would have expected that the strong wrappers _own_ their data, and that the trait has an associated constant of type `ChunkType` for its kind and a method for writing that data. All of the other methods can either live directly on `ChunkType` or aren't really behaviour of that type.